File: 08_Linked_Lists/Reverse_LL/reverse_ll.cpp
// LeetCode Problem 206 - https://leetcode.com/problems/reverse-linked-list/description/
// TC - O(n), SC - O(1)

    ListNode* reverseList(ListNode* head) {
        ListNode* curr = head;
        ListNode* next = NULL;
        ListNode* prev = NULL;
        while(curr != NULL){
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

File: 08_Linked_Lists/Reverse_LL/reverse_ll.java
class reverse_ll {
    public ListNode reverseList(ListNode head) {
        ListNode curr = head;
        ListNode prev = null;
        ListNode next = null;

        while (curr != null) {
            next = curr.next;   // store next node
            curr.next = prev;   // reverse pointer
            prev = curr;        // move prev forward
            curr = next;        // move curr forward
        }

        return prev;  // new head
    }
}
